### PR TITLE
chore: bump Go to 1.17.13

### PR DESCRIPTION
**Pull request by bshaaban**
Fixes all pre Go 1.17.13 related CVEs in Notary.
TODO: update Gorm from github.com/jinzhu/gorm to gorm.io/gorm to take advantage of latest gorm API updates. (requires some refactoring of the gorm API calls)

As part of upgrading to Go 1.17.X, code refactoring was required:
- protobuffs under proto/ and trustmanager/remoteks/ folders were rebuilt
- test make target is now configured to run in Docker to make execution uniform (a new dockerfile was added for this purpose: build.dockerfile)
- all previous linting tools (most of them were outdated) are replaced by staticcheck
- linting is now configured to run in Docker (a new dockerfile was added for this purpose: build.dockerfile)
- some staticcheck code deprecation flags were manually ignored as it will require behavior changes.
  This will require attention. For now original code for these special cases is kept as-is to maintain backward compatibility. See comments added in the following files:
  * cryptoservice/crypto_service.go (line 157)
  * server/server.go (lines 127 and 235)
  * tuf/utils/x509.go (lines 92 and 94)

Signed-off-by: Bahaiddine A <bshaaban@mirantis.com>

### CODE REVIEW

**Comment by justincormack** [comment]
Hi, sorry I merged a gorm update, can you rebase?

**Comment by bshaaban** [comment]
> Hi, sorry I merged a gorm update, can you rebase?

rebased, and reverted go mod upgrade to go 1.17 to skip dependencies updates and minimize changes for now. Any idea why the circleCI action is not running?

**Comment by bshaaban** [comment]
@justincormack, any updates?

**Review by jonnystoten — COMMENTED** [review]
Can we also bump to 1.17 in go.mod?

**Comment by jonnystoten** [comment]
I found the CI to not run on one of my PRs too, the fix seemed to be to rebase a few times until it started running again. :shrug:

**Comment by bshaaban** [comment]
> I found the CI to not run on one of my PRs too, the fix seemed to be to rebase a few times until it started running again. 🤷

yeah, it seems to be running after a push now, thanks.

**Comment by justincormack** [comment]
Can we rebase this now?

**Comment by bshaaban** [comment]
> Can we also bump to 1.17 in go.mod?

originally i have tried that, but then i had to `go mod vendor` and i had over 1M line changes. After rebasing over the gorm bump change, i decided to minimize my change to only bump the images' Go versions.

But since tools/x/sys require an update, i had to update the vendor folder anyway, i just pushed an update with the linting fixes. I'll see if it's worthwhile bumping go.mod's go version to 1.17 ( hopefully it's not as big this time)

**Comment by bshaaban** [comment]
It seems `golint` is outdated starting with Go 1.17, the preferred linter is [staticcheck](https://staticcheck.io/docs/getting-started/), installed using: `go install honnef.co/go/tools/cmd/staticcheck@latest`

There's a ton of linter errors with `staticcheck`. I don't think it's right to push this in this PR. Can we comment out the linter commands for now?

When replacing `golint` with `staticcheck` in the first lint command (all other subsequent linter commands commented out, and probably not needed anymore) like so:
`	@test -z "$(shell find . -type f -name "*.go" -not -path "./vendor/*" -not -name "*.pb.*" -exec golint {} \; | tee /dev/stderr)"`
to
`	@test -z "$(shell find . -type f -name "*.go" -not -path "./vendor/*" -not -name "*.pb.*" -exec staticcheck {} \; | tee /dev/stderr)"`,

I get the following output:
```
% make lint 
xargs: go: terminated with signal 13; aborting
pkcs11 import was not found anywhere without a build tag, yay
cmd/notary-signer/dump_unsupported.go:5:6: func protect is unused (U1000)
cmd/notary-signer/config.go:43:6: func parseSignerConfig is unused (U1000)
cmd/notary-signer/config.go:85:6: func getEnv is unused (U1000)
cmd/notary-signer/config.go:91:6: func passphraseRetriever is unused (U1000)
cmd/notary-signer/config.go:103:6: func setUpCryptoservices is unused (U1000)
cmd/notary-signer/config.go:138:16: error strings should not be capitalized (ST1005)
cmd/notary-signer/config.go:182:6: func getDefaultAlias is unused (U1000)
cmd/notary-signer/config.go:197:6: func setupGRPCServer is unused (U1000)
cmd/notary-signer/config.go:231:6: func getAddrAndTLSConfig is unused (U1000)
cmd/notary-signer/config.go:245:6: func bootstrap is unused (U1000)
cmd/notary-signer/dump_linux.go:10:14: Prctl not declared by package unix (compile)
cmd/notary-signer/dump_linux.go:10:25: PR_SET_DUMPABLE not declared by package unix (compile)
cmd/notary-signer/main.go:64:13: undeclared name: protect (compile)
...
... very long list
...
client/client_test.go:3297:12: undeclared name: NewFileCachedRepository (compile)
client/client_test.go:3376:8: undeclared name: applyTargetsChange (compile)
client/client_test.go:3395:44: undeclared name: repository (compile)
client/client_test.go:3453:21: undeclared name: applyTargetsChange (compile)
client/client_test.go:3454:21: undeclared name: applyTargetsChange (compile)
client/client_test.go:3466:21: undeclared name: applyTargetsChange (compile)
client/client_test.go:3489:21: undeclared name: applyTargetsChange (compile)
client/client_test.go:3490:21: undeclared name: applyTargetsChange (compile)
client/client_test.go:3496:21: undeclared name: applyTargetsChange (compile)
client/client_test.go:3527:12: undeclared name: newCreateDelegationChange (compile)
client/client_test.go:3529:26: undeclared name: tufDir (compile)
client/client_test.go:3532:2: undeclared name: addChange (compile)
client/client_test.go:3536:21: undeclared name: applyTargetsChange (compile)
client/client_test.go:3567:21: undeclared name: applyTargetsChange (compile)
client/client_test.go:3568:21: undeclared name: applyTargetsChange (compile)
client/client_test.go:3580:12: undeclared name: newUpdateDelegationChange (compile)
client/client_test.go:3582:26: undeclared name: tufDir (compile)
client/client_test.go:3585:2: undeclared name: addChange (compile)
client/client_test.go:3589:21: undeclared name: applyTargetsChange (compile)
client/client_test.go:3600:44: undeclared name: repository (compile)
client/client_test.go:3612:12: undeclared name: NewFileCachedRepository (compile)
client/client_test.go:3708:8: undeclared name: DeleteTrustData (compile)
client/client_test.go:3721:42: undeclared name: tufDir (compile)
client/client_test.go:3772:8: undeclared name: DeleteTrustData (compile)
client/client_test.go:3785:42: undeclared name: tufDir (compile)
client/client_test.go:3980:8: undeclared name: applyChangelist (compile)
client/client_test.go:4006:8: undeclared name: applyChangelist (compile)
client/client_test.go:4024:33: undeclared name: TargetSignedStruct (compile)
client/client_test.go:4119:58: undeclared name: TargetSignedStruct (compile)
client/client_test.go:4119:130: undeclared name: TargetSignedStruct (compile)
client/helpers.go:70:10: undeclared name: witnessTargets (compile)
+ lint: golint, go vet, go fmt, gocycle, misspell, ineffassign
# golint
make: *** [lint] Error 1
```

**Comment by bshaaban** [comment]
I'm currently replacing the lint target to use staticcheck and run it under docker. I'll also update go.mod to use go 1.17 and update all dependencies. Hopefully I'll get this out today.

**Comment by bshaaban** [comment]
@jonnystoten @justincormack any reason why CircleCI job does not start anymore? i tried pushing an empty update several times to trigger it but with no luck.

**Comment by bshaaban** [comment]
> It looks like the CI is running on a machine image that went EOL on May 31. I'll have a look at upgrading it.

oh lucky me!

Thanks for letting me know. Much appreciated.

**Comment by jonnystoten** [comment]
Should be sorted with #1645!

**Comment by justincormack** [comment]
ok I merged #1645 - can you rebase and it should run the tests....

**Comment by bshaaban** [comment]
weird, `make test` is passing for me locally (macOS). I don't get any of these errors.

**Comment by justincormack** [comment]
Hmm, weird...

**Comment by jonnystoten** [comment]
I can reproduce this locally. `make test` works for me too but the CI runs:

1. `docker build --progress=plain -t notary_client .`
2. `docker run --rm -e NOTARY_BUILDTAGS --env-file buildscripts/env.list --user notary notary_client bash -c "make ci && codecov"`

The first step is where this fails. The `Dockerfile` in the root has been modified to run `make all` which also runs the tests, and these tests fail as a result. (Locally I also get another test failure:
```
#11 126.9 --- FAIL: TestSetupCryptoServicesRethinkDBStoreConnectionFails (10.00s)
#11 126.9     main_test.go:139: 
#11 126.9         	Error Trace:	main_test.go:139
#11 126.9         	Error:      	"error starting rethinkdb driver: rethinkdb: dial tcp: lookup host on 192.168.65.5:53: read udp 172.17.0.2:52474->192.168.65.5:53: i/o timeout" does not contain "no such host"
#11 126.9         	Test:       	TestSetupCryptoServicesRethinkDBStoreConnectionFails
#11 126.9 FAIL
```
The check in the error text here can probably be removed, it's quite brittle.)

Do we need the changes to `Dockerfile`? They seem a little weird because the first build stage isn't used by the following ones. I wonder if the new stuff at the bottom should be in another Dockerfile if it's needed for something else? Another issue is that the `notary` user isn't created in the resulting image so the `--user notary` flag in the CI step will fail.

If I remove the new stages from `Dockerfile` then running `make ci` explicitly in step 2 above does succeed. I haven't looked into what is causing the difference yet.

**Comment by bshaaban** [comment]
Thank you so much for the analysis @jonnystoten

I've moved the new additions added to Dockerfile into a separate file (`build.dockerfile`) ..

**Comment by jonnystoten** [comment]
Hmm and now the CI isn't running again :sob:

Could you try re-pushing it @bshaaban ?

**Comment by jonnystoten** [comment]
> CircleCI doesn't seem to have Docker installed

This is because the `make ci` step is already running in a container, so we shouldn't try to spin up more containers in there. We could have two make steps `lint` and `lint-docker` or something like that, where `lint-docker` runs `lint` inside a container. WDYT?

> Not sure what's the source of this AWS timeout error
This is probably just a transient error and the job needs to be run again. I don't have permission to run it again, perhaps another re-push is the easiest way.

**Comment by bshaaban** [comment]
> > CircleCI doesn't seem to have Docker installed
> 
> This is because the `make ci` step is already running in a container, so we shouldn't try to spin up more containers in there. We could have two make steps `lint` and `lint-docker` or something like that, where `lint-docker` runs `lint` inside a container. WDYT?

sure, i created separate docker targets for lint and test and kept the originals as-is

> 
> > Not sure what's the source of this AWS timeout error
>  This is probably just a transient error and the job needs to be run again. I don't have permission to run it again, perhaps another re-push is the easiest way.

it looks like job 03 failed again, i'll push another empty change to re trigger the build in a bit

**Comment by bshaaban** [comment]
finally :)

happiness :)

File: Makefile [hunk 1]
```diff
@@ -84,33 +84,43 @@ ${PREFIX}/bin/static/notary:
 	@go build -tags "${NOTARY_BUILDTAGS} netgo" -o $@ ${GO_LDFLAGS_STATIC} ./cmd/notary
 endif
 
+ifeq (, $(shell which staticcheck))
+STATICCHECK_BIN := $(GOBIN)/staticcheck
+$(STATICCHECK_BIN):
+	@echo "+ $@"
+	GOFLAGS="-mod=mod" go install honnef.co/go/tools/cmd/staticcheck@latest
```
Comment (bshaaban) [inline Makefile:91]: `GOFLAGS="-mod=mod"` is necessary here because, most probably, circelCI is set with `-mod=vendor`

**Comment by bshaaban** [comment]
any update on getting this reviewed/merged?

**Comment by bshaaban** [comment]
bumped golang 1.17.10 to 1.17.13 in latest update

**Review by jonnystoten — COMMENTED** [review]
Looks good, just a few comments below

File: utils/http_test.go [hunk 2]
```diff
@@ -82,9 +82,9 @@ func TestWrapWithCacheHeaderNon200Response(t *testing.T) {
 		rw := httptest.NewRecorder()
 		wrapped.ServeHTTP(rw, req)
 
-		require.Equal(t, "", rw.HeaderMap.Get("Cache-Control"))
-		require.Equal(t, "", rw.HeaderMap.Get("Last-Modified"))
-		require.Equal(t, "", rw.HeaderMap.Get("Pragma"))
+		require.Equal(t, "", rw.Header().Get("Cache-Control"))
+		require.Equal(t, "", rw.Header().Get("Last-Modified"))
+		require.Equal(t, "", rw.Header().Get("Pragma"))
```
[the remark this answers is not in the document]
Comment (bshaaban) [reply utils/http_test.go:87]: With the go bump, the `result` variable is of `httptest.ResponseRecorder` type. Its HeaderMap field is now deprecated and the only way we can access the headers is via `Header()` member function. There is no Header field. I can instead create a variable `headers := rw.Header()` and use that variable instead.

[the remark this answers is not in the document]
Comment (jonnystoten) [reply utils/http_test.go]: I don't think that's right - the `rw` variable is of type `httptest.ResponseRecorder`. If you call `Result()` on it, you get an `*http.Response` which has the `Header` field.

Apologies I meant to link [the docs for `ResponseRecorder.Header`](https://pkg.go.dev/net/http/httptest#ResponseRecorder.Header) in the last comment:

> It returns the response headers to mutate within a handler. To test the headers that were written after a handler completes, use the Result method and see the returned Response value's Header.

[the remark this answers is not in the document]
Comment (jonnystoten) [reply utils/http_test.go]: Could you please do the same throughout this file? Then I think this is good to go :heart:

[the remark this answers is not in the document]
Comment (bshaaban) [reply utils/http_test.go]: updated, thanks!

**Review by jonnystoten — APPROVED** [review]

**Review by justincormack — APPROVED** [review]

**Comment by bshaaban** [comment]
great news! we can now focus on bumping go to 1.18.7 / 1.19.2 as a next step?

**Comment by stefan-zh** [comment]
Thank you for upgrading Notary to Go 1.17 and sorry for commenting on a merged pull request. Is there any reason why in `go.mod` we are splitting the dependencies in 2 `require()` blocks: https://github.com/notaryproject/notary/pull/1641/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R34

We are currently trying to pull these changes into our company's fork of Notary and we noticed this.

**Comment by bshaaban** [comment]
> Thank you for upgrading Notary to Go 1.17 and sorry for commenting on a merged pull request. Is there any reason why in `go.mod` we are splitting the dependencies in 2 `require()` blocks: https://github.com/notaryproject/notary/pull/1641/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R34
> 
> We are currently trying to pull these changes into our company's fork of Notary and we noticed this.

there's no specific reason, it must come from `go mod tidy`, they should be combined (if really needed in a future pr or when upgrading to 1.18/1.19)

**Comment by stefan-zh** [comment]
@bshaaban actually, it turns out it's a feature of Go 1.17 to allow for pruning and lazy loading: https://stackoverflow.com/a/69433550/9698467 
So the 2 blocks should be left as is, and shouldn't be combined. Something new for me to learn.

**Pull request merged**